### PR TITLE
chore: denom traces migration handler

### PR DESCRIPTION
**Pull request by charleenfei**
## Description

 closes: #1542 
---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: modules/apps/transfer/keeper/migrations.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/ibc-go/v4/modules/apps/transfer/types"
```
Comment (crodriguezvega) [inline modules/apps/transfer/keeper/migrations.go:5]: Please separate imports :)

File: modules/apps/transfer/keeper/migrations.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+			}
+			iterErr = err
+			return false
+		})
+
+	// replace the outdated traces with the new trace information
+	for _, nt := range newTraces {
+		m.keeper.SetDenomTrace(ctx, nt)
+	}
+	return iterErr
```
Comment (crodriguezvega) [inline modules/apps/transfer/keeper/migrations.go:47]: If a non-`nil` error is returned here, will the changes that may have happened for some denom traces be rolled back?

File: modules/apps/transfer/keeper/migrations.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+			newTrace := types.ParseDenomTrace(dt.GetFullDenomPath())
+			err := newTrace.Validate()
+			if err != nil {
+				iterErr = err
+				return true
+			}
+			if !equalTraces(newTrace, dt) {
+				newTraces = append(newTraces, newTrace)
+			}
+			iterErr = err
```
Comment (crodriguezvega) [inline modules/apps/transfer/keeper/migrations.go:39]: Nit: I guess it doesn't hurt to have it, but I think that you could remove this assignment, since if all denom traces validate successfully `iterErr` will be `nil` because it was never initialized, right?

**Review by crodriguezvega — COMMENTED** [review]
Nice work, @charleenfei!

Should we add changelog entry and also an explanation in the v3-to-v4 migration docs?

File: docs/migrations/v3-to-v4.md [hunk 4]
```diff
@@ -18,7 +18,27 @@ No genesis or in-place migrations required when upgrading from v1 or v2 of ibc-g
...
+
+This incorrect trace information must be corrected when the chain does upgrade to fully supporting denominations with slashes.
+
+If the chain will migrate to supporting base denoms with slashes, it must set the appropriate params during the execution of the upgrade handler in `app.go`: 
+```go
+app.UpgradeKeeper.SetUpgradeHandler("MigrateTraces",
+    func(ctx sdk.Context, _ upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
+        // bump the consensus version of the IBC transfer module 
+        fromVM[ibctransfertypes.ModuleName] = ibctransfermodule.ConsensusVersion()
+
```
Comment (ValarDragon) [inline docs/migrations/v3-to-v4.md:37]: we shouldn't have to do this right? Because the fromVm version would be the old one, and the live code one is `ibctransfermodule.ConsensusVersion()`, hence the `app.mm.RunMigrations` will detect the difference, update the stored version manager, and run the migration right

**Review by ValarDragon — APPROVED** [review]
This is awesome!!

File: docs/migrations/v3-to-v4.md (hunk 4 above)
Comment (charleenfei) [reply docs/migrations/v3-to-v4.md:37]: aha, got it. i wasnt aware that `app.mm.RunMigrations` would make this redundant, and wanted to make the version bumping explicit. but I can remove this line!

File: docs/migrations/v3-to-v4.md [hunk 5]
```diff
@@ -18,7 +18,25 @@ No genesis or in-place migrations required when upgrading from v1 or v2 of ibc-g
 
 ## Chains
 
-- No relevant changes were made in this release.
+# Migrating from not supporing base denoms with slashes to supporting base denoms with slashes
+
+This is necessary when chains are upgrading from a version that does not support base denoms with slashes (e.g. v3.0.0) to a version that does (e.g. v3.1.0). All versions of ibc-go smaller than v1.5.0 for the v1.x release line, v2.3.0 for the v2.x release line, and v3.1.0 for the v3.x release line do *NOT** support IBC token transfers of coins whose base denoms contain slashes. Therefore the in-place of genesis migration described in this document are required when upgrading.
```
Comment (crodriguezvega) [inline docs/migrations/v3-to-v4.md:23]: I am thinking that maybe we can reduce the text here and just make a reference to the migration that was documented in v3.1.0, v2.3.0 and v1.5.0 and say that this is an improved, simplified way to run the same migration. What about something like this?

_As part of [v1.5.0](https://github.com/cosmos/ibc-go/releases/tag/v1.5.0), [v2.3.0](https://github.com/cosmos/ibc-go/releases/tag/v2.3.0) and [v3.1.0](https://github.com/cosmos/ibc-go/releases/tag/v3.1.0) some [migration handler code sample was documented](https://github.com/cosmos/ibc-go/blob/main/docs/migrations/support-denoms-with-slashes.md#upgrade-proposal) that needs to run in order to correct the trace information of coins transferred using ICS20 whose base denom contains slashes.
Based on feedback from the community we add now an improved solution to run the same migration that does not require copying a large piece of code over from the migration document, but instead requires only adding a one-line upgrade handler._

What do you think? Do you agree with writing something like this? Please change whatever you see fit.

**Review by crodriguezvega — COMMENTED** [review]
The is a great improvement, @charleenfei!

Question: should we also update the [existing sample in the separate migration doc](https://github.com/cosmos/ibc-go/blob/main/docs/migrations/support-denoms-with-slashes.md#upgrade-proposal) so that we can backport this change to v2 and v3? Then, for example, we could release v3.2.0 and v2.4.0 and chains would benefit from using the simpler upgrade handler. What do you think?

File: modules/apps/transfer/keeper/migrations.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	keeper Keeper
+}
+
+// NewMigrator returns a new Migrator.
+func NewMigrator(keeper Keeper) Migrator {
+	return Migrator{keeper: keeper}
+}
+
+// This migration will update the denom
+func (m Migrator) MigrateTraces(ctx sdk.Context) error {
```
Comment (damiannolan) [inline modules/apps/transfer/keeper/migrations.go:20]: nit: godocs should start with function name. e.g. "MigrateTraces will update the denom traces stored in state"

File: modules/apps/transfer/keeper/migrations.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	m.keeper.IterateDenomTraces(ctx,
+		func(dt types.DenomTrace) (stop bool) {
+			// check if the new way of splitting FullDenom
+			// into Trace and BaseDenom passes validation and
+			// is the same as the current DenomTrace.
+			// If it isn't then store the new DenomTrace in the list of new traces.
+			newTrace := types.ParseDenomTrace(dt.GetFullDenomPath())
+			err := newTrace.Validate()
+			if err != nil {
+				iterErr = err
```
[the remark this answers is not in the document]
Comment (damiannolan) [reply modules/apps/transfer/keeper/migrations.go:34]: Maybe it should just panic and then have the chain devs export a snapshot and manually address the situation? What do you think?

File: modules/apps/transfer/keeper/migrations_test.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
+package keeper_test
+
+import (
+	"fmt"
+
+	transferkeeper "github.com/cosmos/ibc-go/v4/modules/apps/transfer/keeper"
+	transfertypes "github.com/cosmos/ibc-go/v4/modules/apps/transfer/types"
+)
+
+func (suite *KeeperTestSuite) TestMigrator_MigrateTraces() {
```
Comment (damiannolan) [inline modules/apps/transfer/keeper/migrations_test.go:10]: nit: I think you can remove the underscore in the test func name. This is more like a java-esque style naming convention for tests. I think `TestMigratorMigrateTraces` is good and follows Go conventions

File: modules/apps/transfer/keeper/migrations_test.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(fmt.Sprintf("case %s", tc.msg), func() {
+			suite.SetupTest() // reset
+			tc.malleate()     // explicitly set up denom traces
+			migrator := transferkeeper.NewMigrator(suite.chainA.GetSimApp().TransferKeeper)
+			err := migrator.MigrateTraces(suite.chainA.GetContext())
+			suite.Require().NoError(err)
```
Comment (damiannolan) [inline modules/apps/transfer/keeper/migrations_test.go:102]: nit: I think some kind of logical spacing is nice for readability, but not required! :) 

```suggestion
			suite.SetupTest() // reset
			
			tc.malleate() // explicitly set up denom traces
			
			migrator := transferkeeper.NewMigrator(suite.chainA.GetSimApp().TransferKeeper)
			err := migrator.MigrateTraces(suite.chainA.GetContext())
			suite.Require().NoError(err)
```

**Review by damiannolan — COMMENTED** [review]
Awesome work, looks more or less good to go, just left a few nits and suggestion re. malformed traces

File: modules/apps/transfer/keeper/migrations_test.go [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+						BaseDenom: "customport/channel-7/uatom", Path: "transfer/channel-0/transfer/channel-1",
+					})
+			},
+			transfertypes.Traces{
+				{
+					BaseDenom: "customport/channel-7/uatom", Path: "transfer/channel-0/transfer/channel-1",
+				},
+			},
+		},
+	}
```
[the remark this answers is not in the document]
Comment (charleenfei) [reply modules/apps/transfer/keeper/migrations_test.go:94]: see below comment

File: modules/apps/transfer/keeper/migrations.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (damiannolan) [reply modules/apps/transfer/keeper/migrations.go:34]: In which case, couldn't we return the error immediately? Instead of continuing to do unnecessary work... I'm assuming that the upgrade handler and `RunMigrations` will [likely bubble up the error](https://github.com/cosmos/cosmos-sdk/blob/main/types/module/module.go#L444-L447) and revert any changes so I think it seems unnecessary to continue parsing traces if we encounter an error for some unforeseen reason!

[the remark this answers is not in the document]
Comment (charleenfei) [reply modules/apps/transfer/keeper/migrations.go:34]: we do stop the parsing and then return the error in the current code, is there another way to return the error that you think would make more sense?

[the remark this answers is not in the document]
Comment (damiannolan) [reply modules/apps/transfer/keeper/migrations.go:34]: Ah sorry, I missed the `return true` underneath here due to the thread of comments!

**Review by damiannolan — APPROVED** [review]
Great work! 🚀

File: docs/migrations/v3-to-v4.md [hunk 11]
```diff
@@ -18,7 +18,27 @@ No genesis or in-place migrations required when upgrading from v1 or v2 of ibc-g
...
+As part of [v1.5.0](https://github.com/cosmos/ibc-go/releases/tag/v1.5.0), [v2.3.0](https://github.com/cosmos/ibc-go/releases/tag/v2.3.0) and [v3.1.0](https://github.com/cosmos/ibc-go/releases/tag/v3.1.0) some [migration handler code sample was documented](https://github.com/cosmos/ibc-go/blob/main/docs/migrations/support-denoms-with-slashes.md#upgrade-proposal) that needs to run in order to correct the trace information of coins transferred using ICS20 whose base denom contains slashes.
+
+Based on feedback from the community we add now an improved solution to run the same migration that does not require copying a large piece of code over from the migration document, but instead requires only adding a one-line upgrade handler.
+
+If the chain will migrate to supporting base denoms with slashes, it must set the appropriate params during the execution of the upgrade handler in `app.go`: 
+```go
+app.UpgradeKeeper.SetUpgradeHandler("MigrateTraces",
+    func(ctx sdk.Context, _ upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
+        // transfer module consensus version has been bumped to 2
+        return app.mm.RunMigrations(ctx, app.configurator, fromVM)
```
Comment (chatton) [inline docs/migrations/v3-to-v4.md:32]: Where/how is `RunMigrations` defined? Does this get run at the sdk level as a callback?

This one liner is very nice compared to previous implementation btw, nice job!

File: modules/apps/transfer/keeper/migrations.go [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+	"github.com/cosmos/ibc-go/v4/modules/apps/transfer/types"
+)
+
+// Migrator is a struct for handling in-place store migrations.
+type Migrator struct {
+	keeper Keeper
+}
+
+// NewMigrator returns a new Migrator.
+func NewMigrator(keeper Keeper) Migrator {
```
Comment (chatton) [inline modules/apps/transfer/keeper/migrations.go:15]: is the intention that all transfer related migrations will happen here?

File: modules/apps/transfer/keeper/migrations_test.go [hunk 13]
```diff
@@ -0,0 +1,110 @@
+package keeper_test
+
+import (
+	"fmt"
+
+	transferkeeper "github.com/cosmos/ibc-go/v4/modules/apps/transfer/keeper"
+	transfertypes "github.com/cosmos/ibc-go/v4/modules/apps/transfer/types"
+)
+
+func (suite *KeeperTestSuite) TestMigratorMigrateTraces() {
```
Comment (chatton) [inline modules/apps/transfer/keeper/migrations_test.go:10]: 🥇

**Review by chatton — APPROVED** [review]
This looks great, amazing test coverage! I just had a few question about how the migrations actually run.

File: docs/migrations/v3-to-v4.md (hunk 11 above)
Comment (colin-axner) [reply docs/migrations/v3-to-v4.md:32]: see [docs](https://docs.cosmos.network/main/core/upgrade.html). Core IBC has also had its "module version" [incremented](https://github.com/cosmos/ibc-go/blob/9776639b1a13811191c875c73e185efee577468f/modules/core/module.go#L140) previously

File: modules/apps/transfer/types/trace.go [hunk 14]
```diff
@@ -71,6 +71,9 @@ func (dt DenomTrace) GetFullDenomPath() string {
 	if dt.Path == "" {
 		return dt.BaseDenom
 	}
+	if dt.BaseDenom == "" {
+		return dt.Path
+	}
```
Comment (colin-axner) [inline modules/apps/transfer/types/trace.go:76]: why was this added? I think in this situation, there should be a panic

File: modules/apps/transfer/keeper/migrations_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (colin-axner) [reply modules/apps/transfer/keeper/migrations_test.go:94]: I agree this situation is unlikely. It can still be useful to add a test case to ensure there is an if statement checking that iterErr is non nil (I have seen those missing if statements go unnoticed before)

File: modules/apps/transfer/types/trace.go (hunk 14 above)
Comment (charleenfei) [reply modules/apps/transfer/types/trace.go:76]: i added it bc in the case that the incorrectly parsed BaseDenom is an empty string for whatever reason, the `GetPrefix()` method will actually append an additional `/` onto the end and the BaseDenom will end up looking like ie `erc/0x00000000000etc/` at the conclusion of this logic. I can also set a panic, which do you think makes more sense?

File: modules/apps/transfer/keeper/migrations.go (hunk 12 above)
Comment (charleenfei) [reply modules/apps/transfer/keeper/migrations.go:15]: yes -- the cosmos sdk has all migrations in a [separate directory](https://github.com/cosmos/cosmos-sdk/tree/main/x/bank/migrations) but imo that might be overkill for us. wdyt?

File: modules/apps/transfer/types/trace.go (hunk 14 above)
Comment (colin-axner) [reply modules/apps/transfer/types/trace.go:76]: I think we can cleanup a lot of this code when we look into refactoring transfer and improving upon it. For now, I think it'd be best to remove this check and ensure during migrations that the previously set denom trace passes the `Validate()` call. If it doesn't, we should probably panic

Comment (charleenfei) [reply modules/apps/transfer/types/trace.go:76]: but i dont think some of the previous set denom traces would pass the Validate() call... ie something like ```transfertypes.DenomTrace{
BaseDenom: "1", Path: "transfer/channel-0/gamm/pool",
})```

wouldnt this make the point of the migration code redundant if we are trying to `Validate` before setting the new trace?

File: modules/apps/transfer/keeper/migrations.go (hunk 12 above)
Comment (chatton) [reply modules/apps/transfer/keeper/migrations.go:15]: I think for us it makes sense to have them here, we can worry about additional directories if we end up with a large number of migrations.

File: modules/apps/transfer/types/trace.go (hunk 14 above)
Comment (colin-axner) [reply modules/apps/transfer/types/trace.go:76]: previous invalid denom traces aren't possible in the current code as they would fail on packet [Validation](https://github.com/cosmos/ibc-go/blob/aaccbd193f19f97f8068278141044bf324e6425c/modules/apps/transfer/types/packet.go#L54). The purpose of the migration code is to fix the stored trace info, not validate the denomination. denomTrace.Validate() should pass for all existing denominations

The specific example, "gamm/pool/1", should have failed due to port/channel identifier validation (insufficient length) when being received or sent in a packet

Comment (charleenfei) [reply modules/apps/transfer/types/trace.go:76]: got it, that makes sense!

**Review by AdityaSripal — APPROVED** [review]
Nice work!
